pages/instrumenting.md

```diff
@@ -175,14 +175,75 @@ BigBrotherBird.
 
 Ids are encoded as [hex strings](https://github.com/twitter/finagle/blob/master/finagle-core/src/main/scala/com/twitter/finagle/tracing/Id.scala):
 
-* X-B3-TraceId: 64 encoded bits
-* X-B3-SpanId: 64 encoded bits
-* X-B3-ParentSpanId: 64 encoded bits
-* X-B3-Sampled: Boolean (either "1" or "0")
-* X-B3-Flags: a Long
+* X-B3-TraceId: 64 lower-hex encoded bits (required)
+* X-B3-SpanId: 64 lower-hex encoded bits (required)
+* X-B3-ParentSpanId: 64 lower-hex encoded bits (absent on root span)
+* X-B3-Sampled: Boolean (either "1" or "0", can be absent)
+* X-B3-Flags: "1" means debug (can be absent)
 
 **Thrift Tracing**
 
 Finagle clients and servers negotate whether they can handle extra information
 in the header of the thrift message when a connection is established. Once
 negotiated trace data is packed into the front of each thrift message.
+
+Timestamps and duration
+=====
+
+Span recording is when timing information or metadata is structured and reported
+to zipkin. One of the most important parts of this process is appropriately
+recording timestamps and duration.
+
+** Timestamps are microseconds **
+All Zipkin timestamps are in epoch microseconds (not milliseconds). This value
+should use the most precise measurement available. For example, `clock_gettime`
+or syncing nanoTime against a tick of currentTimeMillis. If none of this works,
```

Comment:
I don't think extracting epoch microseconds by using a tick of currentTimeMillis works. I tried to do something along these lines and the clock drift/correction that the JVM does with currentTimeMillis means your epoch microseconds gets out of whack extremely quickly.

If someone has a valid solution I'd love to see it, but I don't think one exists unfortunately. :(

What *does* work is just multiplying epoch millis by 1000 like you said, and then using nanoTime to calculate the duration. The start timestamp loses a little resolution due to syncing to a tick of currentTimeMillis, but the duration is at least microsecond accurate.